controller/gateway/controller_reconciler_utils.go

```diff
@@ -637,19 +639,38 @@ func (g *gatewayConditionsAndListenersAwareT) setConflicted() {
 // Gateway Programmed status to true.
 // It also sets the listeners Programmed condition by setting the underlying
 // Listener Programmed status to true.
-func (g *gatewayConditionsAndListenersAwareT) setProgrammedAndListenersConditions() {
+func (g *gatewayConditionsAndListenersAwareT) setProgrammedAndListenersConditions(ctx context.Context, c client.Client, logger logr.Logger) {
 	k8sutils.SetProgrammed(g)
 
-	for i := range g.Spec.Listeners {
+	for i := range g.Status.Listeners {
+		listener := &g.Status.Listeners[i]
 		programmedCondition := metav1.Condition{
 			Type:               string(gatewayv1.ListenerConditionProgrammed),
 			Status:             metav1.ConditionTrue,
 			Reason:             string(gatewayv1.ListenerReasonProgrammed),
 			ObservedGeneration: g.GetGeneration(),
 			LastTransitionTime: metav1.Now(),
 		}
-		listenerStatus := listenerConditionsAware(&g.Status.Listeners[i])
+		listenerStatus := listenerConditionsAware(listener)
 		k8sutils.SetCondition(programmedCondition, listenerStatus)
+
+		var routesCount int32
+		if lo.ContainsBy(listener.SupportedKinds, func(gvk gatewayv1.RouteGroupKind) bool {
+			return gvk.Kind == "HTTPRoute" &&
+				gvk.Group != nil && *gvk.Group == gatewayv1.Group(gatewayv1.GroupVersion.Group)
+		}) {
+			httpRoutes, err := gatewayutils.ListHTTPRoutesForGateway(ctx, c, g.Gateway)
+			if err != nil {
+				log.Error(logger, err, "failed to list HTTPRoutes for Gateway when counting AttachedRoutes", g.Gateway)
+				continue
+			}
+
+			routesCount += int32(len(httpRoutes))
+		}
+
+		// NOTE: Count other types of routes when they are supported.
+
+		listenerStatus.AttachedRoutes = routesCount
```

Comment:
Can we extract this code here into a proper function? Also, I think it is more relevant to have it called from the `SetAccepted` function, as it does not really have anything to do with the Programmed condition. What do you think?

Reply:
Done. PTAL